### PR TITLE
Use jar to unzip weird files

### PR DESCRIPTION
**Pull request by olliethomas**
- Added class and folder for audio datsets base, can be instantiated.
- Added placeholder file for concrete ecoacoustics custom dataset.
- Added torchaudio to dependencies.
- Base audio dataset __init__ mostly implemented.
- Added dependencies for audio backend, accomodating for diferent OS types.
- Added utility functionality to infer audio backend based on OS.
- Added module designation file.
- Implemented sample loading and waveform transformation application.
- Added string representation function and import * override.
- Black formatter.
- Implemented basic init function and also unzip check function.
- Implemented basic test for ecoacoustics dataset.
- Added test for already downloaded data.
- Implemented simple metadata extraction function.
- Implemented target/label loading and into metadata extraction.
- Added test for metadata extraction.
- Added decorator.
- Improved naming of import.
- Renamed dataset and all subsequent refernces to it.
- Simplified label extraction by joining the labels and metadata DFs.
- Simplfied return statement.
- Renamed var that matches pyhton built-in.
- Replace string with constant.
- Disambiguated var names.
- Add slow test decorator.
- Incorporated pandas string functions for metadata extraction.
- Create processed audio file if it does not already exist.
- Implemented spectrogram segmenting function.
- Implemented preprocessing of waveforms into variable size spectrogram segments.
- Adapted metadata extraction process to work for pre-processed data rather than .wav files.
- Implemented downloading of data and indices files.
- Fixed downloading issues.
- Updated downloading to use pytorch utils.
- use jar to unzip the dataset files

### CODE REVIEW

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 1]
```diff
@@ -128,14 +171,44 @@ def _download_files(self) -> None:
...
+        remove_finished: bool = False,
+    ) -> str | None:
+        download_root = os.path.expanduser(download_root)
+        if not filename:
+            filename = os.path.basename(url)
+
+        download_url(url, download_root, filename, md5)
+
+        archive = os.path.join(download_root, filename)
+        print(f"Extracting {archive}")
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:201]:
```suggestion
        self.log(f"Extracting {archive}")
```

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 2]
```diff
@@ -128,14 +171,44 @@ def _download_files(self) -> None:
...
+        for finfo in self._URL_MD5_LIST:
+            if not self._check_integrity(finfo.filename, finfo.md5):
+                self.download_and_extract_archive_jar(
+                    url=finfo.url, download_root=str(self.base_dir), md5=finfo.md5
+                )
+
+    def download_and_extract_archive_jar(
+        self,
+        url: str,
+        download_root: str,
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:189]:
```suggestion
        download_root: str,
```
just replace this with`self._base_dir`?

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 3]
```diff
@@ -89,6 +124,14 @@ def __init__(
 
         super().__init__(x=x, y=y, s=s, transform=transform, audio_dir=self.base_dir)
 
+    def _check_integrity(self, filename: str, md5: str) -> bool:
+        root = self.root
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:128]: why?

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 4]
```diff
@@ -89,6 +124,14 @@ def __init__(
...
+        fpath = os.path.join(root, self.base_dir, filename)
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:129]:
```suggestion
        fpath = self.base_dir / filename
```

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 5]
```diff
@@ -128,14 +171,40 @@ def _download_files(self) -> None:
...
+
+        for finfo in self._URL_MD5_LIST:
+            if not self._check_integrity(finfo.filename, finfo.md5):
+                self.download_and_extract_archive_jar(url=finfo.url, md5=finfo.md5)
+
+    def download_and_extract_archive_jar(
+        self,
+        url: str,
+        md5: str,
+        filename: str | None = None,
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:188]:
```suggestion
        file_info: ZenodoInfo
```

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 6]
```diff
@@ -89,6 +124,14 @@ def __init__(
 
         super().__init__(x=x, y=y, s=s, transform=transform, audio_dir=self.base_dir)
 
+    def _check_integrity(self, filename: str, md5: str) -> bool:
+        root = self.root
+        fpath = os.path.join(root, self.base_dir, filename)
+        if not check_integrity(fpath, md5):
+            return False
+        print(f"{filename} already downloaded.")
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:132]:
```suggestion
        self.log(f"{filename} already downloaded.")
```

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 7]
```diff
@@ -89,6 +124,14 @@ def __init__(
 
         super().__init__(x=x, y=y, s=s, transform=transform, audio_dir=self.base_dir)
 
+    def _check_integrity(self, filename: str, md5: str) -> bool:
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:127]:
```suggestion
    def _check_integrity(self, filename: str, *, md5: str) -> bool:
```

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 8]
```diff
@@ -29,6 +37,12 @@
 Extension = Literal[".pt", ".wav"]
 
 
+class ZenodoInfo(NamedTuple):
+    filename: str
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:41]:
```suggestion
    filename: str | None
```

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 9]
```diff
@@ -128,14 +171,40 @@ def _download_files(self) -> None:
...
+        if not archive_type:
+            return _decompress(
+                str(archive),
+                os.path.join(self.base_dir, filename.replace(suffix, "")),
+                remove_finished=remove_finished,
+            )
 
-        for url in urls:
-            download_and_extract_archive(url, str(self.base_dir))
+        subprocess.run(f"jar -xvf {archive}", shell=True, check=True, cwd=self.base_dir)
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:207]: Can we have some kind of error-handling here (e.g. informing the user that jar needs to be installed)?

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 10]
```diff
@@ -128,14 +171,40 @@ def _download_files(self) -> None:
...
+        download_url(url, str(self.base_dir), filename, md5)
+
+        archive = self.base_dir / filename
+        self.log(f"Extracting {archive}")
+
+        suffix, archive_type, compression = _detect_file_type(str(archive))
+        if not archive_type:
+            return _decompress(
+                str(archive),
+                os.path.join(self.base_dir, filename.replace(suffix, "")),
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:203]:
```suggestion
                self.base_dir / filename.with_suffix(""),
```

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 11]
```diff
@@ -172,39 +171,39 @@ def _download_files(self) -> None:
...
-
-        subprocess.run(f"jar -xvf {archive}", shell=True, check=True, cwd=self.base_dir)
+            return None
+
+        try:
+            subprocess.run(f"jar -xvf {archive}", shell=True, check=True, cwd=self.base_dir)
+        except:
+            self.log(
+                "Tried to extract malformed .zip file using Java."
+                "However, there was a porblem. Is Java on your system path?"
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:205]:
```suggestion
                "However, there was a problem. Is Java on your system path?"
```

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 12]
```diff
@@ -128,14 +170,40 @@ def _download_files(self) -> None:
...
+        if not archive_type:
+            _ = _decompress(
+                str(archive),
+                str((self.base_dir / finfo.filename).with_suffix("")),
+                remove_finished=remove_finished,
+            )
+            return
+
+        try:
+            subprocess.run(f"jar -xvf {archive}", check=True, cwd=self.base_dir)
```
Comment (thomkeh) [inline bolts/data/datasets/audio/ecoacoustics.py:201]:
```suggestion
            subprocess.run(["jar", "-xvf", str(archive)], check=True, cwd=self.base_dir)
```

Comment (thomkeh) [reply bolts/data/datasets/audio/ecoacoustics.py:201]: if you don't have `shell=True` you need to do the splitting on spaces manually (normally the shell does that for you).

**Comment by MylesBartlett** [comment]
Does this work now?

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 13]
```diff
@@ -257,18 +258,20 @@ def _preprocess_audio(
         if not self._processed_audio_dir.exists():
             mkdir(self._processed_audio_dir)
 
-        waveform_paths = list(self.base_dir.glob("**/*.wav"))
+        waveform_paths = [
+            path for path in self.base_dir.glob("**/*.wav") if "MACOSX" not in str(path)
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:262]:
```suggestion
            path for path in self.base_dir.glob("**/*.wav") if "MACOSX" not in str(path)
```
Checking each file-name is rather inefficient; why not just glob the EC/UK_BIRD subdirectories directly?

File: bolts/data/datasets/audio/ecoacoustics.py [hunk 14]
```diff
@@ -257,18 +258,20 @@ def _preprocess_audio(
...
+        ]
 
         tform = transform(n_fft=n_freq_bins, hop_length=hop_len)
-        for path in waveform_paths:
+        for path in tqdm(waveform_paths, desc="Preprocessing"):
```
Comment (MylesBartlett) [inline bolts/data/datasets/audio/ecoacoustics.py:266]:
```suggestion
        for path in tqdm(total=waveform_paths, desc="Preprocessing"):
```

File: bolts/data/datasets/audio/ecoacoustics.py (hunk 13 above)
Comment (MylesBartlett) [reply bolts/data/datasets/audio/ecoacoustics.py:262]: Alternatively, you could glob with an exclusion pattern:   
`image_paths.extend(self._base_dir.glob(f"**/[!__]*.{ext}"))`

Comment (MylesBartlett) [reply bolts/data/datasets/audio/ecoacoustics.py:262]: Alternatively, the MACOS directory could be deleted upon download

**Review by MylesBartlett — APPROVED** [review]

**Pull request merged**